code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServiceImpl.java

```diff
@@ -381,9 +386,104 @@ public long getServiceNetworkId(Service service) {
         return networkId;
     }
 
-    protected String getServiceInstanceName(Service service, int order) {
+    protected List<? extends String> getServiceInstancesNames(Service service, int requestedCount, boolean toAdd) {
+        List<Integer> usedIds = getServiceInstanceUsedOrderIds(service);
+        if (toAdd) {
+            Collections.sort(usedIds);
+        } else {
+            Collections.sort(usedIds, Collections.reverseOrder());
+        }
+        List<String> serviceNames = new ArrayList<>();
+        if (toAdd) {
+            int finalCount = 0;
+            // in situation when service with scale=3 has one container missing (it got destroyed outside of the
+            // service)
+            // and container names don't reflect the order, we should pick the least available number that is <=order
+            for (int i = 1; i < requestedCount + 1; i++) {
+                if (finalCount < requestedCount) {
+                    if (!usedIds.contains(i)) {
+                        serviceNames.add(generateServiceInstanceName(service, i));
+                        finalCount++;
+                        usedIds.add(i);
+                    }
+                }
+            }
+            // only after we got all "gap" names, get the rest
+            Collections.sort(usedIds);
```

Comment:
I'm a little confused about this part.  At this point, usedIds should be complete (original + any holes filled including potentially new ones if requestedCount is higher than before).  But finalCount will just be the # of holes filled (by holes, i'm once again including potentially new ones).